rfcs/2020-08-21-3092-apache-metrics-source.md

```diff
@@ -0,0 +1,235 @@
+```
+
+I'll use this to generate the following metrics:
+
+- `apache.uptime_seconds` (counter)
+- `apache.total_accesses` (counter; extended)
+- `apache.total_kilobytes` (counter; extended)
+- `apache.total_duration` (counter; extended)
+- `apache.cpu_user` (gauge; extended)
+- `apache.cpu_system` (gauge; extended)
+- `apache.cpu_children_user` (gauge; extended)
+- `apache.cpu_children_system` (gauge; extended)
+- `apache.cpu_load` (gauge; extended)
+- `apache.requests_per_second` (gauge; extended)
+- `apache.bytes_per_second` (gauge; extended)
+- `apache.bytes_per_request` (gauge; extended)
+- `apache.duration_per_request` (gauge; extended)
+- `apache.busy_workers` (gauge)
+- `apache.idle_workers` (gauge)
+- `apache.processes` (gauge)
+- `apache.stopping` (gauge)
+- `apache.conns_total` (gauge)
+- `apache.conns_async_writing` (gauge)
+- `apache.conns_async_keepalive` (gauge)
+- `apache.conns_async_closing` (gauge)
+- `apache.scoreboard_waiting` (gauge)
+- `apache.scoreboard_starting` (gauge)
+- `apache.scoreboard_reading` (gauge)
+- `apache.scoreboard_sending` (gauge)
+- `apache.scoreboard_keepalive` (gauge)
+- `apache.scoreboard_dnslookup` (gauge)
+- `apache.scoreboard_closing` (gauge)
+- `apache.scoreboard_logging` (gauge)
+- `apache.scoreboard_finishing` (gauge)
+- `apache.scoreboard_idle_cleanup` (gauge)
+- `apache.scoreboard_open` (gauge)
+
+Metrics labeled `extended` are only available if `ExtendedStatus` is enabled
+for Apache. This is the default in newer versions (>= 2.4; released 2012), but
+purportedly [increases CPU
+load](https://www.datadoghq.com/blog/collect-apache-performance-metrics/#a-note-about-extendedstatus)
+so some users may turn it off. If it is off, they simply won't have those
+metrics published.
+
+I figure we probably don't want metrics for:
+
+- Load (should be handled by a `cpu` or similar metrics source)
+
+## Doc-level Proposal
+
+Users will be instructed to setup
+[`mod_status`](https://httpd.apache.org/docs/2.4/mod/mod_status.html) and
+enable
+[`ExtendedStatus`](https://httpd.apache.org/docs/2.4/mod/core.html#extendedstatus).
+
+The following additional source configuration will be added:
+
+```toml
+[sources.my_source_id]
+  type = "apache" # required
+  endpoint = "http://localhost/server-status?auto" # required, default
+  scrape_interval_secs = 15 # optional, default, seconds
+```
+
+Some possible configuration improvements we could add in the future would be:
+
+- `response_timeout`; to cap request lengths
+- `tls`: settings to allow setting specific chains of trust and client certs
+
+But I chose to leave those out for now given the prometheous source doesn't
```

Comment:
I think the big missing piece here is Basic auth - a lot of people still use this as security for mod_status. Prom just adds it to the scrape URL.

Reply:
Agreed, I think we could support this in the same way (via the URL), but that it'd be useful to add basic auth options for all HTTP-based client sources. I propose deferring that until after we refactor them as mentioned in "Future Work".

Reply:
Ah - missed that. +1.